service/hub/server/server.go

```diff
@@ -0,0 +1,373 @@
+package server
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"net"
+	"net/http"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/cesanta/glog"
+	"github.com/docker/distribution/registry/auth/token"
+	"github.com/labring/service/hub/api"
+	"github.com/labring/service/hub/auth"
+)
+
+var (
+	hostPortRegex = regexp.MustCompile(`^(?:\[(.+)\]:\d+|([^:]+):\d+)$`)
+	scopeRegex    = regexp.MustCompile(`([a-z0-9]+)(\([a-z0-9]+\))?`)
+)
+
+type AuthServer struct {
+	config         *Config
+	authenticators api.Authenticator
+	authorizers    api.Authorizer
+}
+
+func NewAuthServer(c *Config) (*AuthServer, error) {
+	as := &AuthServer{
+		config:         c,
+		authenticators: auth.NewSealosAuthn(),
+		authorizers:    auth.NewSealosAuthz(),
+	}
+	return as, nil
+}
+
+type AuthRequest struct {
+	RemoteConnAddr string
+	RemoteAddr     string
+	RemoteIP       net.IP
+	User           string
+	Password       api.PasswordString
+	Account        string
+	Service        string
+	Scopes         []AuthScope
+	Labels         api.Labels
+}
+
+type AuthScope struct {
+	Type    string
+	Class   string
+	Name    string
+	Actions []string
+}
+
+type AuthzResult struct {
+	scope            AuthScope
+	autorizedActions []string
+}
+
+func (ar AuthRequest) String() string {
+	return fmt.Sprintf("{%s:%s@%s %s}", ar.User, ar.Password, ar.RemoteAddr, ar.Scopes)
+}
+
+func parseRemoteAddr(ra string) net.IP {
+	hp := hostPortRegex.FindStringSubmatch(ra)
+	if hp != nil {
+		if hp[1] != "" {
+			ra = hp[1]
+		} else if hp[2] != "" {
+			ra = hp[2]
+		}
+	}
+	res := net.ParseIP(ra)
+	return res
+}
+
+func parseScope(scope string) (string, string, error) {
+	parts := scopeRegex.FindStringSubmatch(scope)
+	if parts == nil {
+		return "", "", fmt.Errorf("malformed scope request")
+	}
+
+	switch len(parts) {
+	case 3:
+		return parts[1], "", nil
+	case 4:
+		return parts[1], parts[3], nil
+	default:
+		return "", "", fmt.Errorf("malformed scope request")
+	}
+}
+
+func (as *AuthServer) ParseRequest(req *http.Request) (*AuthRequest, error) {
+	ar := &AuthRequest{RemoteConnAddr: req.RemoteAddr, RemoteAddr: req.RemoteAddr}
+	if as.config.Server.RealIPHeader != "" {
+		hv := req.Header.Get(as.config.Server.RealIPHeader)
+		ips := strings.Split(hv, ",")
+
+		realIPPos := as.config.Server.RealIPPos
+		if realIPPos < 0 {
+			realIPPos = len(ips) + realIPPos
+			if realIPPos < 0 {
+				realIPPos = 0
+			}
+		}
+
+		ar.RemoteAddr = strings.TrimSpace(ips[realIPPos])
+		glog.V(3).Infof("Conn ip %s, %s: %s, addr: %s", ar.RemoteAddr, as.config.Server.RealIPHeader, hv, ar.RemoteAddr)
```

Comment:
The logs are not uniform, using both log and glog?

Reply:
ok, I will fix it and use glog.